File: parser/parser.cpp
#include "token.h"
#include "parser.h"


#include <errno.h>
#include <exception>
#include <stdexcept>
#include <stdio.h>
#include <iostream>
#include <sstream>
#include <cstring>

using namespace std;



ScriptParser::ScriptParser(const string &file_name){
		currentfile = file_name;
        s.start(currentfile);
        lookahead = s.nextToken();
        last=nullptr;
}

void ScriptParser::stop(){
}

void ScriptParser::match(token_t c) {
	cout << lookahead->print() << endl;
    if (lookahead->code==c) {
        last = lookahead;
        lookahead=s.nextToken();
    } else error(c,lookahead->code);
}

// Skip ahead one token without matching
void ScriptParser::skip() {
    last = lookahead;
    lookahead=s.nextToken();
}

token_t ScriptParser::nextToken() {
    return lookahead->code;
}

int64_t ScriptParser::nextTokenAsInteger() {
    return lookahead->intValue();
}

int64_t ScriptParser::lastTokenAsInteger() {
    return last->intValue();
}

string ScriptParser::nextTokenAsString() {
    return lookahead->stringValue();
}

string ScriptParser::lastTokenAsString() {
    return last->stringValue();
}


double ScriptParser::nextTokenAsFloat(){
	return lookahead->floatValue();
}

double ScriptParser::lastTokenAsFloat(){
	return last->floatValue();
}


Range ScriptParser::nextTokenAsRange() {
    return  lookahead->rangeValue();
}

Range ScriptParser::lastTokenAsRange() {
    return  last->rangeValue();
}


void ScriptParser::error(token_t expect, token_t found){
    cerr << ": PARSE ERROR:  Expecting " << token_name.at(expect) << " but found " << token_name.at(found) << endl;
    cerr << "On line " <<  s.line << " in file " << currentfile << endl;
    exit(1);
}

void ScriptParser::error(const string &errstr){
    cerr << "Parse error: " << errstr << endl;
    cerr << "On line " <<  s.line << " in file " << currentfile << endl;
    exit(1);
}

File: Go/parsing/parser.go
package parsing

import (
	"fmt"
	"io"
)

type ParseError struct {
	Line    int
	Message string
}

func (e *ParseError) Error() string {
	return fmt.Sprintf("Parse error at line %d: %s", e.Line, e.Message)
}

type Parser struct {
	scanner   *Scanner
	lookahead *Token
	lastToken *Token
}

func NewParser(r io.Reader) *Parser {
	p := &Parser{
		scanner: NewScanner(r),
	}
	p.lookahead = p.scanner.NextToken()
	return p
}

func (p *Parser) Match(expected TokenType) error {
	if p.lookahead.Type != expected {
		return &ParseError{
			Line:    p.lookahead.Line,
			Message: fmt.Sprintf("expected %s, got %s", tokenTypeNames[expected], p.lookahead.String()),
		}
	}
	p.Skip()
	return nil
}

func (p *Parser) Skip() {
	p.lastToken = p.lookahead
	p.lookahead = p.scanner.NextToken()
}

func (p *Parser) NextToken() *Token {
	return p.lookahead
}

func (p *Parser) LastTokenAsString() string {
	if p.lastToken != nil {
		return p.lastToken.StrValue
	}
	return ""
}

func (p *Parser) LastTokenAsInteger() int64 {
	if p.lastToken != nil {
		return p.lastToken.IntValue
	}
	return 0
}

func (p *Parser) LastTokenAsFloat() float64 {
	if p.lastToken != nil {
		return p.lastToken.FloatValue
	}
	return 0.0
}

func (p *Parser) LastTokenAsRange() Range {
	if p.lastToken != nil {
		return p.lastToken.RangeValue
	}
	return Range{}
}

// Deprecated - scanner now skips newlines
func (p *Parser) SkipNewlines() {
	// No-op - kept for compatibility
}